geospaas_processing/copiers.py

```diff
@@ -42,31 +43,44 @@ def write_flag_file(type_in_flag_file, source_path, dataset, destination_filenam
             flag_file.write(string_to_write)
 
     def file_or_symlink_copy(self, source_paths, dataset):
-        """copy the file or a symlink of the file of dataset based on its stored local address in
-        the database."""
+        """copy the file or a symlink of the file/folder of dataset based on its stored local
+        address in the database."""
         for source_path in source_paths:
-            if os.path.isfile(source_path.uri):
-                destination_filename = os.path.join(
-                    self._destination_path, os.path.basename(source_path.uri))
+            self.source_is_file = os.path.isfile(source_path.uri)
+            self.source_is_folder = os.path.isdir(source_path.uri)
+            destination_file_or_folder_name = os.path.join(
+                self._destination_path, os.path.basename(source_path.uri))
+            if (self.source_is_file or self.source_is_folder):
                 # below if condition prevents "shutil.copy" or "os.symlink" from replacing the file
-                # in the destination in a repetitive manner.
-                if not os.path.isfile(destination_filename) or not os.path.islink(
-                        destination_filename):
-                    if self._link_request:
-                        os.symlink(src=source_path.uri, dst=destination_filename)
-                    else:
-                        shutil.copy(src=source_path.uri, dst=self._destination_path)
-                    if self._flag_file_request:
-                        self.write_flag_file(self._type_in_flag_file,
-                                             source_path, dataset, destination_filename)
+                # or folder in the destination in a repetitive manner.
+                if (not (os.path.isfile(destination_file_or_folder_name)
+                         or os.path.isdir(destination_file_or_folder_name))
+                    or not os.path.islink(destination_file_or_folder_name)):
+                    self.copy_item(source_path, destination_file_or_folder_name, dataset)
                 else:
                     LOGGER.debug(
-                        "For dataset with id = %s, there is already a symlink or a file with the "
-                        + "same name in the destination folder.", dataset.id)
+                        "For dataset with id = %s, there is already a symlink or a file or a folder"
+                        + "with the same name in the destination folder.", dataset.id)
```

Comment:
That should be at least a warning, if not an error message.

Reply:
It can be made shorter too, something like: "Failed to copy dataset %s: the destination path already exists."

Reply:
Done